src/QsCompiler/Transformations/FunctorGeneration.cs

```diff
@@ -148,6 +152,125 @@ public override QsStatementKind OnConjugation(QsConjugation stm)
+                    this.SharedState.AdditionalStatements.Clear();
+                    if (!(transformed.Statement is QsStatementKind.QsExpressionStatement expr &&
+                        expr.Item.Expression == ExpressionKind.UnitValue))
+                    {
+                        // Only add statements that are not free-floating Unit, which could have
+                        // been left behind by expression transformation.
+                        statements.Add(transformed);
+                    }
+                }
+                return new QsScope(statements.ToImmutableArray(), scope.KnownSymbols);
+            }
+        }
+
+        public class ExpressionTransformation
+        : ExpressionTransformation<TransformationsState>
+        {
+            public ExpressionTransformation(SyntaxTreeTransformation<TransformationsState> parent)
+            : base(parent, TransformationOptions.Default)
+            {
+            }
+
+            public ExpressionTransformation(TransformationsState parent)
+            : base(parent, TransformationOptions.Default)
+            {
+            }
+
+            public override TypedExpression OnTypedExpression(TypedExpression ex)
+            {
+                this.SharedState.CurrentExpression.Push(ex);
+                var newEx = base.OnTypedExpression(ex);
+                this.SharedState.CurrentExpression.Pop();
+                return newEx;
+            }
+        }
+
+        public class ExpressionKindTransformation
+        : ExpressionKindTransformation<TransformationsState>
+        {
+            public ExpressionKindTransformation(SyntaxTreeTransformation<TransformationsState> parent)
+            : base(parent)
+            {
+            }
+
+            public ExpressionKindTransformation(TransformationsState parent)
+            : base(parent)
+            {
+            }
+
+            public override ExpressionKind OnOperationCall(TypedExpression method, TypedExpression arg)
+            {
+                // An operation in an adjoint scope must return Unit, so extract the operation to be an
+                // an additional statement and then replace it with Unit.
+                var curExpression = this.SharedState.CurrentExpression.Peek();
+                this.SharedState.AdditionalStatements.Add(
+                    new QsStatement(
+                        QsStatementKind.NewQsExpressionStatement(
+                            new TypedExpression(
+                                base.OnOperationCall(method, arg),
+                                curExpression.TypeParameterResolutions.Select(x => Tuple.Create(x.Key.Item1, x.Key.Item2, x.Value)).ToImmutableArray(),
+                                curExpression.ResolvedType,
+                                curExpression.InferredInformation,
+                                curExpression.Range)),
+                        LocalDeclarations.Empty,
+                        QsNullable<QsLocation>.Null,
```

Comment:
nit: Could you keep track of and fill out the statement location the same way you did for the expression location? Keeping location information around for as long as possible can help transformations down the line that my want/need to use them.

Reply:
Good idea! I'm working on that now...